contrib/Motif/clients/xmpostit/xmpostit.man
.TH XPOSTIT 1 "2 January 1989" RIACS/X11R3
.SH NAME
xmpostit \- X window system (Motif-like) Post-it\(rg notes
.SH SYNOPSIS
.B xmpostit
[
.I \-toolkitoptions ...
] [
.I \-options ...
]
.SH DESCRIPTION
.I Xmpostit
provides a mechanism for manipulating on-screen Post-it\(rg notes.
All three sizes of Post-it notes may be displayed,
edited,
and saved to disk files or stashed in the main window as buttons.
In addition,
any on-screen note may be resized to any arbitrary size either when it is
created or at a later time.  This size will be retained if the note is "saved."
.PP
When
.I xmpostit
is first invoked,
it creates a small window.
This is the control window for
.IR xmpostit .
Additionally,
if any notes were stored in the save directory (see below),
these will become buttons on the control window.  The label on the button 
is the name of the note.
Each Post-it note,
when created, prompts for a name to use and 
then makes a window for the note.
.I Xmpostit
is controlled using the mouse buttons and a menu.
The Post-it notes are controlled using buttons and the scroll bars.
.SH OPTIONS
.PP
.I Xmpostit
accepts all of the standard X Toolkit command line options along with the
additional options listed below.
.TP 8
.BI \-bs " number"
Specifies the size of the character buffer in which to save each Post-it
note.
A note may not contain more than this number of characters,
including newlines.
The default is 1024.
.TP 8
.BI \-dir " path"
Specifies the path name of the directory in which to store saved Post-it notes.
If
.I path
begins with a `/' it is taken as-is;
if it does not begin with a `/',
the path is taken relative to the home directory.
The named directory must already exist if notes are to be saved in it.
The default is
.IR \&.postitnotes .
.TP 8
.B \-sv
Enables save-on-exit.
When specified,
this option tells
.I xmpostit
to save all Post-it notes to disk files when exiting.
If not specified,
notes will only be saved to disk files when saved manually by the user.
.SH "X DEFAULTS"
.PP
.I Xmpostit
understands all of the core X Toolkit resource names and classes as well
as those listed below,
which are all of class
.BR XmPostit .
.TP 8
.B bufSize
Specifies the size of the character buffer used to store a Post-it note.
The default is 1024.
.TP 8
.B noteDir
Specifies the path name of the directory in which to store saved notes.
The default is
.IR \&.postitnotes .
.TP 8
.B saveNotes
Controls the state of the save-on-exit option.
The default is
.BR false .
.TP 8
.B scrollBar
Controls placing scroll bars on Post-it notes.
The default is
.BR false .
.TP 8
.B showPlaid
displays the 3M-like plaid background found typical of paper
postit notes.  The default is True.
.PP
Because of the large amount of widget nesting in
.I xmpostit ,
the following constructions are recommended for setting resources on
the various widgets:
.in 8
.sp
.nf
*XmPostit.\fIresource\fP

*XmPostit.plaid*\fIresource\fP

*XmPostit.Confirm*\fIresource\fP
*XmPostit.Confirm*confirm*\fIresource\fP
*XmPostit.Confirm*cancel*\fIresource\fP

*XmPostit.Menu*\fIresource\fP

*XmPostit.PostItNote*\fIresource\fP
*XmPostit.PostItNote*save*\fIresource\fP
*XmPostit.PostItNote*erase*\fIresource\fP
*XmPostit.PostItNote*destroy*\fIresource\fP
*XmPostit.PostItNote*text*\fIresource\fP
.in 0
.sp
.fi
.SH "THE CONTROL WINDOW"
.PP
.I Xmpostit
allows two operations to be performed from its control window.
Pressing mouse button one on a button in the 
.I xmpostit
main window causes the note for that button to appear on the screen.
Pressing mouse button three in the control window raises the
.I xmpostit
menu.
.SH "THE MENU"
.PP
The
.I xmpostit
menu provides the following selections:
.TP 8
.B "Create 1.5x2 Note"
Create a new Post-it note,
1.5 inches tall by 2 inches wide.
.TP 8
.B "Create 3x3 Note"
Create a new Post-it note,
3 inches square.
.TP 8
.B "Create 3x5 Note"
Create a new Post-it note,
3 inches tall by 5 inches wide.
.TP 8
.B "Raise All Notes"
Raise all Post-it notes to the top.
.TP 8
.B "Lower All Notes"
Lower all Post-it notes to the bottom.
.TP 8
.B "Save All Notes"
Save all Post-it notes to disk files in the save directory.
.TP 8
.B "Choose Printer"
Change or specify the printer for the "Hardcopy" button.  The
default is $PRINTER.
.TP 8
.B Exit
Exit
.I xmpostit .
If the
.B \-sv
command line option was given,
or the
.B saveNotes
resource is
.B true,
all Post-it notes will be saved to disk first.
.PP
To select an item from the menu,
move the mouse cursor to that item and press the left mouse button.
.SH "THE POST-IT NOTE"
.PP
Each Post-it note is made up of two parts:
a scrolling text window where the text of the note is stored,
and a row of five buttons labeled ``Save'',
``Erase'', ``Destroy'', ``Hardcopy'', and ``Stash''.
.PP
To enter text into a Post-it note,
simply move the mouse cursor into the text window and start typing.
Since the text window is actually a Motif Text widget,
all the Motif Text widget translations including cutting and 
pasting are accepted.
.PP
After entering text in the Post-it note,
you may wish to save the note in a disk file.
This way,
if the machine goes down,
or if you exit
.IR xmpostit ,
the Post-it note can be restored when you restart
.IR xmpostit .
To save the note to a disk file,
move the mouse cursor to the button labeled ``save'' and press the left
mouse button.
The note will be saved as the file ``note\fIn\fP'' in your save directory,
where
.I n
is some sequence number.
.BR \s-1NOTE\s0 :
it is important to remember that until you have pressed the ``save'' button,
the data stored in the Post-it note is not saved on disk and can be lost if
the system goes down or
.I xmpostit
exits.
.PP
To erase the entire contents of the text window,
you can press the left mouse button in the ``erase'' button.
This will bring up a confirmation window,
which has two buttons labeled ``confirm'' and ``cancel''.
If you press the ``confirm'' button,
the text will be erased.
If you press the ``cancel'' button,
the operation is canceled,
and nothing will happen.
.BR \s-1NOTE\s0 :
erasing the text in the window does not affect any contents of the note
you have saved on disk unless you press the ``save'' button again.
.PP
To destroy a Post-it note,
getting rid of its window on the screen and the disk file it is saved in,
move the mouse cursor to the ``destroy'' button and press the left mouse
button.
This will bring up a confirmation window as described above.
If you confirm the operation,
the Post-it note will disappear from the screen and the disk file it was
saved in will be deleted.
.PP
To send the current contents of a Post-it note to a printer, click mouse
button one on ``hardcopy''.  If no printer is set (and $PRINTER is not
set) xmpostit will ring the terminal bell.  Otherwise the current contents
of the note, whether or not saved, will be sent to the selected printer.
To select or change the printer, click on ``Choose Printer'' by popping
up the main xmpostit menu.
.PP
To "stash" the note as a button on the main xmpostit window, click the 
mouse button one on ``Stash.''  This will make the button of the note's
name active and unmap the note from the screen.  The note can also be 
iconified using the currently running window manager's iconify capability.
.SH "SEE ALSO"
X(1)
.SH BUGS
.PP
The sizes of the Post-it notes are only as acurrate as the display
dimension and resolution returned by the server.
On the Sun server and possibly others,
this means they aren't all that accurate.
.PP
Without a color display for canary yellow notes and red plaid,
the aesthetic value of
.I xmpostit
cannot be fully appreciated.  In such a case,
.I showPlaid
should be set False.
.SH AUTHOR
Original author of xpostit:
David A. Curry, Research Institute for Advanced Computer Science
.PP
Motif implementation and design:
David Mischel, Genentech
.sp
``Post-it'' and the plaid design are registered trademarks of 3M.


contrib/Motif/clients/xmpostit/xmpostit.c
/* based on xpostit.c
 *
 * David A. Curry
 * Research Institute for Advanced Computer Science
 * Mail Stop 230-5
 * NASA Ames Research Center
 * Moffett Field, CA 94086
 * davy@riacs.edu
 *
 *
 * Copyright 1989, 1990 Genentech, Incorporated
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Genentech not be used in advertising or
 * publicity pertaining to distribution of the software without specific,
 * written prior permission.  Genentech makes no representations about the
 * suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * GENENTECH DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS, IN NO
 * EVENT SHALL GENENTECH BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF
 * USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 * PERFORMANCE OF THIS SOFTWARE. 
 *
 * Author:  David Mischel - Genentech, Inc.
 */

#include <signal.h>
#include <stdio.h>
#include <Xm/Xm.h>
#include <X11/StringDefs.h>
#include <X11/Intrinsic.h>
#include <X11/Shell.h>
#include <X11/cursorfont.h>
#include <Xm/PushBG.h>
#include <Xm/RowColumn.h>
#include <Xm/DrawingA.h>
#include <Xm/ScrolledW.h>
#include <Xm/SelectioB.h>
#include <Xm/MwmUtil.h>

void	menuSelection();
void	setBackground();

#include "xmpostit.h"
#include "plaid.bm"

#define MENUCREATE_1P5x2	0
#define MENUCREATE_3x3		1
#define MENUCREATE_3x5		2
#define MENURAISEALL		3
#define MENULOWERALL		4
#define MENUSAVEALL		5
#define MENUPRINTER		6
#define MENUEXIT		7

void	PrinterNameCB();
void	PrinterNameCancelCB();

AppRes	app_res;		/* xpostit application resources	*/
XtAppContext appcontext;	/* application context			*/
Widget	toplevel;		/* top level application shell widget	*/
Screen	*screen;		/* pointer to the screen of the display	*/
Display	*display;		/* pointer to the display we're on	*/
Widget	menuwidget;
Widget		rcwidget;
char printername[MAXNAME +1];	/* device for "hardcopy" button */

/*
 * Command line options and the resources they set.
 */
static XrmOptionDescRec options[] = {
    { "-bs",	".bufSize",		XrmoptionSepArg,	NULL },
    { "-dir",	".noteDir",		XrmoptionSepArg,	NULL },
    { "-sv",	".saveNotes",		XrmoptionNoArg,		"true" }
};

/*
 * Fallback resources.
 */
static String fallback_resources[] = {
    "*Command*font: -adobe-helvetica-bold-r-normal--11-*-*-*-*-*-*-*",
    "*Menu*font:    -adobe-helvetica-medium-r-normal--12-*-*-*-*-*-*-*",
    "*Text*font:    -b&h-lucidatypewriter-medium-r-normal-*-12-*-*-*-*-*-*-*",
    NULL
};
/*
 * Resources we maintain besides those maintained by the toolkit.
 */
static XtResource resources[] = {
#define offset(field)   XtOffset(AppResPtr,field)
    { "bufSize", "BufSize", XtRInt, sizeof(int),
      offset(buf_size), XtRImmediate, (caddr_t) DefaultBufSize },
    { "noteDir", "NoteDir", XtRString, sizeof(String),
      offset(note_dir), XtRString, DefaultNoteDir },
    { "saveNotes", "SaveNotes", XtRBoolean, sizeof(Boolean),
      offset(save_notes), XtRImmediate, (caddr_t) False },
    { "showPlaid", "ShowPlaid", XtRBoolean, sizeof(Boolean),
      offset(show_plaid), XtRImmediate, (caddr_t) True },
    { "scrollOnOverflow", "Scroll", XtRBoolean, sizeof(Boolean),
      offset(scroll_ovf), XtRImmediate, (caddr_t) False }
#undef offset
};

void
main(argc, argv)
char **argv;
int argc;
{
	/*
	 * Initialize the toolkit and create an application shell.
	 */
	toplevel = XtAppInitialize(&appcontext, PostItNoteClass, options,
			XtNumber(options), &argc, argv, fallback_resources,
			NULL, 0);

	display = XtDisplay(toplevel);
	screen = DefaultScreenOfDisplay(display);

   	 /*
   	  * Now get any resources we're interested in.
  	   */
  	XtGetApplicationResources(toplevel, &app_res, resources,
        XtNumber(resources), argv, argc);

	SetNoteDir();
	CreatePlaidWidget();

	XtRealizeWidget(toplevel);
	if (app_res.show_plaid) setBackground();
	XFlush(display);

	XtAppMainLoop(appcontext);
}

static void	HandlePlaidButton();
static Widget scrollW, clipA;

void
CreatePlaidWidget()
{
	Arg args[11];
	register int nargs;
	Cursor menuC;

	strcpy(printername, (char *) getenv("PRINTER"));
	nargs = 0;
    SetArg (XmNscrollBarDisplayPolicy, XmAS_NEEDED);
    SetArg (XmNscrollingPolicy, XmAUTOMATIC);
    scrollW = XmCreateScrolledWindow (toplevel, "scrollW", args, nargs);
	XtManageChild(scrollW);

	nargs = 0;
	SetArg(XmNwidth, DefaultPlaidWidth);
	SetArg(XmNheight, DefaultPlaidHeight);
	SetArg(XmNpacking, XmPACK_COLUMN);
	SetArg(XmNadjustMargin, False);
	SetArg(XmNadjustLast, False);
	SetArg(XmNmarginHeight, 12);
	SetArg(XmNmarginWidth, 12);
	SetArg(XmNentryBorder, 6);
	rcwidget = XmCreateRowColumn(scrollW, "rcwidget", args, nargs);
	XtManageChild(rcwidget);
	menuC = XCreateFontCursor(display, XC_pencil);
	XmSetMenuCursor(display, menuC);

	CreateMenuWidget();

	nargs = 0;
	SetArg(XmNclipWindow, &clipA);
	XtGetValues(scrollW, args, nargs);
	XtAddEventHandler(clipA, ButtonPressMask, False,
		HandlePlaidButton, menuwidget);

	/*
	 * Load the notes the user has saved, and create widgets
	 * for them.
	 */
	LoadSavedNotes();
}

  void
setBackground()
{
	Arg args[11];
	register int nargs;
	int depth;
	Pixel fg, bg;
	Pixmap plaidmap;

	nargs = 0;
	SetArg(XmNforeground, &fg);
	SetArg(XmNbackground, &bg);
	SetArg(XmNdepth, &depth);
	XtGetValues(rcwidget, args, nargs);
	plaidmap = XCreatePixmapFromBitmapData(display, XtWindow(rcwidget),
		plaid_bits, plaid_width, plaid_height, fg, bg, depth);

	/* set the background for the clip area too */
	nargs = 0;
	SetArg(XmNbackgroundPixmap, plaidmap);
	XtSetValues(clipA, args, nargs);
	XtSetValues(rcwidget, args, nargs);
}

/*
 * HandlePlaidButton - even handler for plaid widget
 */
/* ARGSUSED */
  static void
HandlePlaidButton(w, popup, event)
	Widget w, popup;
	XButtonEvent *event;
{
	Arg args[4];
	register int nargs;
	register Position x, y;


	/*
	 * We're only interested in button presses.
	 */
	if (event->type != ButtonPress)
		return;

	/*
	 * Dispatch a function for this button.
	 */
	switch (event->button) {
	case Button3:
		/*
		 * Get position of mouse so we can put
		 * the menu there.
		 */
		x = event->x_root;
		y = event->y_root;

		/*
		 * Move the popup to that location.
		 */
		nargs = 0;
		SetArg(XtNx, x);
		SetArg(XtNy, y);
		XtSetValues(menuwidget, args, nargs);

		/*
		 * Bring up the menu.
		 */
		XmMenuPosition(popup, event);
		XtManageChild(popup);
		break;
	default:
		break;
	}
}

/*
 * CreateMenuWidget - create the widget used for the menu.
 */
void
CreateMenuWidget()
{
	Widget popupBtn[8];
	int popn;
	Arg args[4];
	register int nargs;

	nargs = 0;
	SetArg(XmNwhichButton, (XtArgVal) 3); 
	SetArg(XmNmenuCursor, "pencil");

	menuwidget = XmCreatePopupMenu(rcwidget, "plaidmenu", args, nargs);
	XtAddEventHandler(rcwidget, ButtonPressMask, False,
		HandlePlaidButton, menuwidget);

	popn = 0;
	popupBtn[popn] = XmCreatePushButtonGadget(menuwidget,
		"1.5x3 Note", NULL, 0);
	XtAddCallback(popupBtn[popn], XmNactivateCallback,
		menuSelection, MENUCREATE_1P5x2);
	popn++;

	popupBtn[popn] = XmCreatePushButtonGadget(menuwidget,
		"3x3 Note", NULL, 0);
	XtAddCallback(popupBtn[popn], XmNactivateCallback,
		menuSelection, MENUCREATE_3x3);
	popn++;

	popupBtn[popn] = XmCreatePushButtonGadget(menuwidget,
		"3x5 Note", NULL, 0);
	XtAddCallback(popupBtn[popn], XmNactivateCallback,
		menuSelection, MENUCREATE_3x5);
	popn++;

	popupBtn[popn] = XmCreatePushButtonGadget(menuwidget,
		"Raise Notes", NULL, 0);
	XtAddCallback(popupBtn[popn], XmNactivateCallback,
		menuSelection, MENURAISEALL);
	popn++;

	popupBtn[popn] = XmCreatePushButtonGadget(menuwidget,
		"Lower Notes", NULL, 0);
	XtAddCallback(popupBtn[popn], XmNactivateCallback,
		menuSelection, MENULOWERALL);
	popn++;

	popupBtn[popn] = XmCreatePushButtonGadget(menuwidget,
		"Save Notes", NULL, 0);
	XtAddCallback(popupBtn[popn], XmNactivateCallback,
		menuSelection, MENUSAVEALL);
	popn++;

	popupBtn[popn] = XmCreatePushButtonGadget(menuwidget, "Choose Printer", NULL, 0);
	XtAddCallback(popupBtn[popn], XmNactivateCallback,
		menuSelection, MENUPRINTER);
	popn++;

	popupBtn[popn] = XmCreatePushButtonGadget(menuwidget, "Exit", NULL, 0);
	XtAddCallback(popupBtn[popn], XmNactivateCallback,
		menuSelection, MENUEXIT);
	popn++;

	XtManageChildren(popupBtn, popn);
}

/*
 * menuSelection - callback from list widget to handle a selection.
 */
/* ARGSUSED */
  void
menuSelection(w, client_data, call_data)
caddr_t client_data, call_data;
Widget w;
{
	/*
	 * Dispatch the command.
	 */
	switch ((int) client_data) {
	case MENUCREATE_1P5x2:
		CreateNewNote(PostItNote_1p5x2);
		break;
	case MENUCREATE_3x3:
		CreateNewNote(PostItNote_3x3);
		break;
	case MENUCREATE_3x5:
		CreateNewNote(PostItNote_3x5);
		break;
	case MENURAISEALL:
		RaiseAllNotes();
		break;
	case MENULOWERALL:
		LowerAllNotes();
		break;
	case MENUSAVEALL:
		SaveAllNotes();
		break;
	case MENUPRINTER:
		ChoosePrinter();
		break;
	case MENUEXIT:
		ByeBye();

	}
}

  void
ChoosePrinter()
{
	Arg args[5];
	int nargs;
	Widget twidget;

	nargs = 0;
	SetArg(XmNdialogType, XmDIALOG_PROMPT);
	SetArg(XmNselectionLabelString,
		XmStringCreateLtoR("Name of Printer for Hardcopy",
		XmSTRING_DEFAULT_CHARSET));
	SetArg(XmNtextString, 
		XmStringCreateLtoR(printername, XmSTRING_DEFAULT_CHARSET));
	SetArg(XmNdialogStyle, XmDIALOG_APPLICATION_MODAL);
	SetArg(XmNtextColumns, MAXNAME);
	twidget = XmCreatePromptDialog(rcwidget, "Printer", args, nargs);
	XtAddCallback(twidget, XmNokCallback, PrinterNameCB, NULL);
	XtAddCallback(twidget, XmNcancelCallback, PrinterNameCancelCB, NULL);

    XtUnmanageChild
        (XmSelectionBoxGetChild (twidget, XmDIALOG_HELP_BUTTON));
	
	/* turn off mwm borders */
	nargs = 0;
	SetArg(XmNmwmDecorations, MWM_DECOR_ALL);
	XtSetValues(XtParent(twidget), args, nargs);
	/* limit input text size */
	nargs = 0;
	SetArg(XmNmaxLength, MAXNAME);
	XtSetValues(XmSelectionBoxGetChild(twidget, XmDIALOG_TEXT), args, nargs);

	XtRealizeWidget(twidget);
	XtManageChild(twidget);
}


/* ARGSUSED */
  void
PrinterNameCB(w, client_data, call_data)
	Widget w;
	caddr_t client_data, call_data;
{
	Widget textW;

	textW = XmSelectionBoxGetChild(w, XmDIALOG_TEXT);
	strcpy(printername, XmTextGetString(textW));
	XtDestroyWidget(XtParent(w));
}


/* ARGSUSED */
  void
PrinterNameCancelCB(w, client_data, call_data)
	Widget w;
	caddr_t client_data, call_data;
{
	XtDestroyWidget(XtParent(w));
}

